### PR TITLE
Tranches

### PR DESCRIPTION
**Pull request by akiezun**
VQSR tranche functionality. I had to add the VRAC and VariantDatum (also present on another branch) to make things compile. These are simple classes. @droazen please review

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    public final int numKnown;      //number of known sites in this tranche
+    public final int numNovel;      //number of novel sites in this tranche
+    public final String name;       //Name of the tranche
+    public final VariantRecalibratorArgumentCollection.Mode model;    //this is a SNP VQSR tranche or Indel tranche?
+
+    private final int accessibleTruthSites;
+    private final int callsAtTruthSites;
+
+    public Tranche(double targetTruthSensitivity, double minVQSLod, int numKnown, double knownTiTv, int numNovel, double novelTiTv, int accessibleTruthSites, int callsAtTruthSites, VariantRecalibratorArgumentCollection.Mode model) {
+        this(targetTruthSensitivity, minVQSLod, numKnown, knownTiTv, numNovel, novelTiTv, accessibleTruthSites, callsAtTruthSites, model, "anonymous");
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: Make default tranche name `"anonymous"` a named constant

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    public Tranche(double targetTruthSensitivity, double minVQSLod, int numKnown, double knownTiTv, int numNovel, double novelTiTv, int accessibleTruthSites, int callsAtTruthSites, VariantRecalibratorArgumentCollection.Mode model, String name) {
+        if ( targetTruthSensitivity < 0.0 || targetTruthSensitivity > 100.0) {
+            throw new UserException("Target FDR is unreasonable " + targetTruthSensitivity);
+        }
+
+        if ( numKnown < 0 || numNovel < 0) {
+            throw new GATKException("Invalid tranche - no. variants is < 0 : known " + numKnown + " novel " + numNovel);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: Are you sure that an invalid `targetTruthSensitivity` is the user's fault, while an invalid number of known/novel variants is our fault?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        this.knownTiTv = knownTiTv;
+        this.numKnown = numKnown;
+        this.model = model;
+        this.name = name;
+
+        this.accessibleTruthSites = accessibleTruthSites;
+        this.callsAtTruthSites = callsAtTruthSites;
+    }
+
+    public static final Comparator<Tranche> TRUTH_SENSITIVITY_ORDER = (tranche1, tranche2) -> Double.compare(tranche1.targetTruthSensitivity, tranche2.targetTruthSensitivity);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: Move Comparator declaration to top of class.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+    @Override
+    public String toString() {
+        return String.format("Tranche targetTruthSensitivity=%.2f minVQSLod=%.4f known=(%d @ %.4f) novel=(%d @ %.4f) truthSites(%d accessible, %d called), name=%s]",
+                targetTruthSensitivity, minVQSLod, numKnown, knownTiTv, numNovel, novelTiTv, accessibleTruthSites, callsAtTruthSites, name);
+    }
+
+    private static double getRequiredDouble(Map<String,String> bindings, String key ) {
+        if ( bindings.containsKey(key) ) {
+            return Double.valueOf(bindings.get(key));
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: Wrap the potential `NumberFormatException` within either a `UserException` or `GATKException`, depending on whether or not the data came from user input.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        } else  {
+            throw new UserException.MalformedFile("Malformed tranches file.  Missing required key " + key);
+        }
+    }
+
+    private static double getOptionalDouble(Map<String,String> bindings, String key, double defaultValue ) {
+        return Double.valueOf(bindings.getOrDefault(key, String.valueOf(defaultValue)));
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: As above, throw a `UserException` upon a  `NumberFormatException`

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    private static int getRequiredInteger(Map<String,String> bindings, String key) {
+        if ( bindings.containsKey(key) ) {
+            return Integer.valueOf(bindings.get(key));
+        } else {
+            throw new UserException.MalformedFile("Malformed tranches file.  Missing required key " + key);
+        }
+    }
+
+    private static int getOptionalInteger(Map<String,String> bindings, String key, int defaultValue) {
+        return Integer.valueOf(bindings.getOrDefault(key, String.valueOf(defaultValue)));
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: In these two methods as well -- throw a `UserException` upon a `NumberFormatException`

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+     * Returns a list of tranches, sorted from most to least specific, read in from file f.
+     * @throws java.io.IOException if there are problems reading the file.
+     */
+    public static List<Tranche> readTranches(File f) throws IOException{
+        String[] header = null;
+        List<Tranche> tranches = new ArrayList<>();
+
+        try (XReadLines xrl = new XReadLines(f) ) {
+            for (final String line : xrl) {
+                if (line.startsWith("#")) {  //comment
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: Comment character should be a named constant

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+                    continue;
+                }
+
+                final String[] vals = line.split(",");
+                if (header == null) {  //reading the header
+                    header = vals;
+                    if (header.length != 11) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: Number of expected header line elements (11) should be a named constant

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+                        throw new UserException.MalformedFile(f, "Expected 11 elements in header line " + line);
+                    }
+                } else {
+                    if (header.length != vals.length) {
+                        throw new UserException.MalformedFile(f, "Line had too few/many fields.  Header = " + header.length + " vals " + vals.length + ". The line was: " + line);
+                    }
+
+                    Map<String, String> bindings = new HashMap<>();
+                    for (int i = 0; i < vals.length; i++) bindings.put(header[i], vals[i]);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: Enclose for loop body within braces

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+        public String getName() { return name; }
+
+        public abstract double getThreshold(double tranche);
+        public abstract void calculateRunningMetric(List<VariantDatum> data);
+        public abstract double getRunningMetric(int i);
+    }
+
+    public static final class TruthSensitivityMetric extends SelectionMetric {
+        double[] runningSensitivity;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: Should be private

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+        double metricThreshold = metric.getThreshold(trancheThreshold);
+        int n = data.size();
+        for ( int i = 0; i < n; i++ ) {
+            if ( metric.getRunningMetric(i) >= metricThreshold ) {
+                // we've found the largest group of variants with sensitivity >= our target truth sensitivity
+                Tranche t = trancheOfVariants(data, i, trancheThreshold, model);
+                logger.info(String.format("  Found tranche for %.3f: %.3f threshold starting with variant %d; running score is %.3f ",
+                        trancheThreshold, metricThreshold, i, metric.getRunningMetric(i)));
+                logger.info(String.format("  Tranche is %s", t));
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: Should these be info-level or debug?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+                    numNovel++;
+                    if( datum.isSNP ) {
+                        if ( datum.isTransition ) { novelTi++; } else { novelTv++; }
+                    }
+                }
+            }
+        }
+
+        double knownTiTv = knownTi / Math.max(1.0 * knownTv, 1.0);
+        double novelTiTv = novelTi / Math.max(1.0 * novelTv, 1.0);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: Calculation of TiTv given # of transitions/transversions should be a utility method.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+
+        int accessibleTruthSites = VariantDatum.countCallsAtTruth(data, Double.NEGATIVE_INFINITY);
+        int nCallsAtTruth = VariantDatum.countCallsAtTruth(data, minLod);
+
+        return new Tranche(ts, minLod, numKnown, knownTiTv, numNovel, novelTiTv, accessibleTruthSites, nCallsAtTruth, model);
+    }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: These classes suffer from an almost complete lack of javadoc -- do you want to address that in this branch?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/VariantDatum.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    public boolean atTruthSite;
+    public boolean atTrainingSite;
+    public boolean atAntiTrainingSite;
+    public boolean isTransition;
+    public boolean isSNP;
+    public boolean failingSTDThreshold;
+    public double prior;                 //per-site prior
+    public GenomeLoc loc;
+    public int worstAnnotation;
+    public boolean isAggregate; // this datum was provided to aid in modeling but isn't part of the input callset
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/VariantDatum.java]: Non-final public fields should ideally be either properly encapsulated or made final.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/VariantRecalibratorArgumentCollection.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+/*
+ * A bunch of annotations for VQSR.
+ */
+final class VariantRecalibratorArgumentCollection {
+
+    static public enum Mode {
+        SNP,
+        INDEL,
+        BOTH
+    }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/VariantRecalibratorArgumentCollection.java]: I feel like this exact enum is declared several times across the codebase...

File: src/main/java/org/broadinstitute/hellbender/utils/text/XReadLines.java [hunk 16]
```diff
@@ -24,14 +24,18 @@
...
     private final String commentPrefix;
 
-    public XReadLines(final File filename) throws IOException {
-        this(IOUtils.makeReaderMaybeGzipped(filename), true, null);
+    public XReadLines(final File filename) {
+        this(filename, true);
+    }
+
+    public XReadLines(final File filename, final boolean trimWhitespace) {
+        this(IOUtils.makeReaderMaybeGzipped(filename), trimWhitespace, null);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/text/XReadLines.java]: You touched these constructors, therefore you should write javadoc for them :)

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/TrancheUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    @Test(expectedExceptions = {UserException.MalformedFile.class})
+    public final void readOldFormat() throws IOException {
+        read(EXPECTED_TRANCHES_OLD);
+    }
+
+    public final List<Tranche> read(File f) throws IOException{
+        return Tranche.readTranches(f);
+    }
+
+    private static void assertTranchesAreTheSame(List<Tranche> newFormat, List<Tranche> oldFormat, boolean completeP, boolean includeName) {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/TrancheUnitTest.java]: Does the `Tranche` class perhaps needs an `equals()` method?

**Comment by droazen** [comment]
Review complete

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: I'm just going to change to GATKException for now

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 4 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 8 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 9 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: done. good find

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 10 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 11 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: you're right. debug makes more sense. it's not intended for user consumption

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 12 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: a utility method for dividing `a` by `b`?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 13 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: i'll take it on in the next round.
https://github.com/broadinstitute/hellbender/issues/372

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/VariantDatum.java (hunk 14 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/VariantDatum.java]: yes. but don't review this class in this pull request. The proper one for VariantDatum is coming later .

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/VariantRecalibratorArgumentCollection.java (hunk 15 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/VariantRecalibratorArgumentCollection.java]: can't find any that would be suitable here. SAMPileUpFeature seems close but it has `VariantType { NONE, SNP, INSERTION, DELETION, INDEL }`

File: src/main/java/org/broadinstitute/hellbender/utils/text/XReadLines.java (hunk 16 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/text/XReadLines.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/TrancheUnitTest.java (hunk 17 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/TrancheUnitTest.java]: probably not - we never put them into collections etc. And The test needs a special trick to ignore the name so it'd need a different equals which is not worth it.

**Comment by akiezun** [comment]
review comments addressed. back to @droazen to re-review

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java (hunk 12 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java]: It's a little more than that -- `knownTi / Math.max(1.0 * knownTv, 1.0)`

**Comment by droazen** [comment]
There are (real) test failures -- back to you :)

**Comment by akiezun** [comment]
sorry. a blooper - the name needs to be skipped. back to @droazen

**Comment by droazen** [comment]
:+1: looks good -- rebase/merge at will

**Pull request merged**